configdialog.cpp

```diff
@@ -703,3 +734,57 @@ LRESULT CALLBACK CustomColorDialog(HWND dialog, UINT message, WPARAM wparam, LPA
 
 	return FALSE;
 }
+
+LRESULT CALLBACK ScreenSaverImportExportPalettesDialog(HWND dialog, UINT message, WPARAM wparam, LPARAM lparam) {
+	HWND palettes_input = GetDlgItem(dialog, IDC_PALDLG_IMPORT_EXPORT_PALETTES);
+	switch (message) {
+		case WM_INITDIALOG: {
+			Edit_SetText(
+				palettes_input,
+				(wchar_t *) lparam
+			);
+			return TRUE;
+		} case WM_COMMAND: {
+
+			wchar_t palettes_buffer[1 << 16] { L'\0' };
+			Edit_GetText(
+				palettes_input,
+				palettes_buffer,
+				1 << 16
+			);
+			palettes_buffer[(1 << 16) - 1] = L'\0';
+
+			switch (LOWORD(wparam)) {
+				case IDOK: {
+					auto *palettes = new std::wstring(palettes_buffer);
+					EndDialog(dialog, (INT_PTR) palettes);
+					return TRUE;
+				} case IDCANCEL: {
+					EndDialog(dialog, 0);
+					return TRUE;
+				} case IDC_PALDLG_IMPORT_EXPORT_PALETTES: {
+					switch (HIWORD(wparam)) {
+						case EN_CHANGE: {
+							// Various special characters might break the registry parsing if they are saved.
+							// The only catastrophe I can directly forsee is commas, but may as well just only allow
+							// alphanumerics and a few special characters to be extra safe.
+							static std::wstring valid_chars = L"qwertyuiopasdfghjklzxcvbnmQWERTYUIOPASDFGHJKLZXCVBNM0123456789 -_!?:#;~\r\n";
+							std::wstring palettes = palettes_buffer;
+
+							bool palettes_too_short = palettes.size() < PaletteCustomizeDialog::MinPaletteNameSize + 57;
+							bool palettes_has_invalid_chars = palettes.find_first_not_of(valid_chars) != std::wstring::npos;
+
+							HWND ok_button = GetDlgItem(dialog, IDOK);
+							EnableWindow(ok_button, !palettes_too_short && !palettes_has_invalid_chars);
```

Comment:
I can't tell if this is incomplete or not so I'll just mention it, but this validation pass is going to need a lot more to ensure that the code someone pasted in is actually importable. You will have to check that the code is in _exactly_ the same format as the string produced by `export_palettes`. So, it has a palette name followed by a colon followed by seven hex codes followed by another separator... and so on.

Because that is likely going to be a complex validation, it's definitely worth splitting off into its own function somewhere.

Reply:
Yes and no.
It's complete in the sense that I don't plan on adding more validation _here_ but incomplete in the sense that there _will_ be more validation. It will just be a bit more verbose than graying out a button.